### ISSUE TITLE
Generate code from examples even if no `-->` is found

### ISSUE THREAD

**Issue by edkelly303** [lue-bird/elm-review-documentation-code-snippet#1]
Hi lue-bird!

This is a brilliant elm-review rule, thanks for making it.

I noticed that if you define a function/constant in a documentation code example, the rule will generate the code for that function/constant in the test file. But this only happens if the code example ends with a `-->`.

I have some docs where I provide examples of how to use the API of my package, but it doesn't really make sense to show the reader the value that results from the code, so I don't add a `-->` line. For example:

    myDict = 
        dict int string

Where `dict`, `int` and `string` are functions in my package, and the value of `myDict` is a fairly complex opaque type that contains a function, which the user of the package doesn't need to know about.

It would be really useful if the rule would still generate the code for `myDict` in the test file, because then if I rename one of my API functions but forget to update the name in the code example, I'll get a compilation error in the test file.

What do you think? Would this be possible/desirable? Or maybe I should rethink the way I write my doc examples so that they always end in a `-->`?

**Comment by lue-bird** [comment]
You hint that the usual solution (convert the declaration to a `--:`) is not desirable in your case.

That's indeed an interesting use case that I didn't think existed because module-level declarations are usually always annotated.

I had made the decision to leave snippets without an expectation mark alone because you often want some snippets to not be valid elm
```elm
{-| 

    Http.request { url = "invalid url", ... }

will on execution come back with a `Http.InvalidUrl`
-}
```
```elm
{-| 

    import List.Extra

    [ 1, 2, 3 ]
        |> List.Extra.minimum
        |> Maybe.withDefault ??

Above should not happen, so how can we make it return a value, not a `Maybe`?


-}
```
```elm
{-|

    -- length >= 5
    ArraySized ... (Min (Up5 x_))
    
    -- 2 <= length <= 12
    ArraySized ... (In (Up2 minX_) (Up12 maxX_))

-}
```
```elm
{-| 

    1                ________
                   /         \
    0 ____________/           \_______
       t ->    fade in     fade out

-}
```

Any suggestions on what specifically should be checked and what not are welcome.

**Comment by edkelly303** [comment]
Thanks for this thoughtful response - I had not considered the case of having code examples that are *deliberately* not valid Elm code.

Maybe a good option for my use case would be a separate rule, something like `Review.Documentation.NoInvalidCodeSnippetDeclarations`, which would simply generate a file with all code from doc examples (without necessarily generating tests) and thus allow the package author to check that all their declarations compile successfully. It could potentially also take a list of function names as config so that the package author could include or exclude specific functions from the rule (e.g. if they have some functions whose docs contain code examples that are deliberately intended to be invalid Elm code, like the examples you showed in your comment).

I totally appreciate that this is probably out of scope for what you're doing with this package, but it looks like you've done 90% of the work that would be required to make my idea work, so if you don't want to go in this direction, I might be able to make something myself using your rule as a starting point.

Thanks very much for talking this through with me.

**Comment by lue-bird** [comment]
I'm currently thinking that the easiest option might be to add a new marker like
```elm
dict int string
--: BiDict.Interface ...
```
(the `...` would need to be present somewhere)
This would not check for a specific type at all, only check that above expression compiles.

**Comment by edkelly303** [comment]
Ah, I think I'm confused. Is `--:`, followed by the type of the preceding definition, already an accepted convention in doc examples, or is this a new notation that you are suggesting? I've never noticed this being used in Elm package docs, but maybe I've not been paying enough attention!

**Comment by lue-bird** [comment]
`--:` exists and is mentioned in the rule documentation https://dark.elm.dmy.fr/packages/lue-bird/elm-review-documentation-code-snippet/latest/Review-Documentation-CodeSnippet#type-checks

**Comment by edkelly303** [comment]
Ah, sorry, I missed that completely!

For the package I'm working on (edkelly303/elm-any-type-forms), the types are often horribly complicated, so they'd be a distraction for the reader of the docs if I used `--:`. 

I guess the `...` option would mean I don't have to write out the types in full, so that would reduce the noise, but it would still mean adding an extra line of text to each doc example that doesn't necessarily give the user of the package any useful information. So for my case, I'm not sure it would be worth adding the `...` functionality to the rule.

As I've thought about it more, I actually think the better option in my case is for me to add a `test` function to my API that would allow a user to programmatically initialize a form control, pass some messages to it to update its state, and then validate it and check the output. I could then use this `test` function in my doc examples and use your rule to check everything works correctly. As a bonus, it would also allow users of my package to test their form controls with elm-test more easily.

In summary, I now think the rule is fine as-is - sorry for wasting your time and thanks again for discussing this with me.
